File: src/main/java/lekkit/scev/machine/MouseDevice.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.machine;

/**
 * Abstract HID mouse.
 *
 * <p>Coordinates are virtual-pixel coordinates inside a resolution set via
 * {@link #resolution(int, int)}; relative {@link #move(int, int)} is in
 * device units. Buttons follow the HID bitmask convention:
 * {@code 1 = left, 2 = right, 4 = middle}.
 */
public interface MouseDevice {
    /** Declare the virtual display dimensions for coordinate translation. */
    void resolution(int width, int height);

    /** Place the pointer at an absolute coordinate. */
    void place(int x, int y);

    /** Move the pointer by a relative delta. */
    void move(int dx, int dy);

    /** Press one or more buttons (HID bitmask). */
    void press(byte hidButtons);

    /** Release one or more buttons. */
    void release(byte hidButtons);

    /** Scroll wheel: positive = up, negative = down. */
    void scroll(byte delta);
}

File: src/main/kotlin/lekkit/scev/machine/MouseDevice.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.machine

/**
 * Abstract HID mouse.
 *
 * Coordinates are virtual-pixel coordinates inside a resolution set
 * via [resolution]; relative [move] is in device units. Buttons follow
 * the HID bitmask convention: `1 = left, 2 = right, 4 = middle`.
 */
interface MouseDevice {
    /** Declare the virtual display dimensions for coordinate translation. */
    fun resolution(width: Int, height: Int)

    /** Place the pointer at an absolute coordinate. */
    fun place(x: Int, y: Int)

    /** Move the pointer by a relative delta. */
    fun move(dx: Int, dy: Int)

    /** Press one or more buttons (HID bitmask). */
    fun press(hidButtons: Byte)

    /** Release one or more buttons. */
    fun release(hidButtons: Byte)

    /** Scroll wheel: positive = up, negative = down. */
    fun scroll(delta: Byte)
}
